File: tests/ClaudeToolbar.Core.Tests/Sessions/HooksConfigTests.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using ClaudeToolbar.Core.Sessions;

namespace ClaudeToolbar.Core.Tests.Sessions;

public class HooksConfigTests
{
    private const string Url = "http://127.0.0.1:47831/hook";

    private static JsonObject Obj(string json) => (JsonObject)JsonNode.Parse(json)!;

    [Fact]
    public void UrlUsesThePort() => Assert.Equal(Url, HooksConfig.HookUrl(47831));

    [Fact]
    public void InstallIntoEmptyCreatesEveryEvent()
    {
        var json = HooksConfig.Install("", Url);
        var hooks = (JsonObject)Obj(json)["hooks"]!;
        Assert.Equal(HooksConfig.Events, hooks.Select(kv => kv.Key).ToList());
        foreach (var ev in HooksConfig.Events)
        {
            var handler = (JsonObject)((JsonArray)((JsonObject)((JsonArray)hooks[ev]!)[0]!)["hooks"]!)[0]!;
            Assert.Equal("http", handler["type"]!.GetValue<string>());
            Assert.Equal(Url, handler["url"]!.GetValue<string>());
            Assert.Equal(5, handler["timeout"]!.GetValue<int>());
        }
        Assert.True(HooksConfig.IsInstalled(json, Url));
        Assert.False(HooksConfig.IsInstalled("{}", Url));
        Assert.False(HooksConfig.IsInstalled("", Url));
    }

    [Fact]
    public void InstallIsIdempotentAndPreservesOtherContent()
    {
        const string existing = """
        {
          "model": "opus",
          "hooks": {
            "Stop": [ { "hooks": [ { "type": "command", "command": "say done" } ] } ],
            "PreToolUse": [ { "matcher": "Bash", "hooks": [ { "type": "command", "command": "lint" } ] } ]
          }
        }
        """;
        var once = HooksConfig.Install(existing, Url);
        var twice = HooksConfig.Install(once, Url);
        Assert.Equal(once, twice);
        var root = Obj(twice);
        Assert.Equal("opus", root["model"]!.GetValue<string>());
        var stop = (JsonArray)((JsonObject)root["hooks"]!)["Stop"]!;
        Assert.Equal(2, stop.Count);
        Assert.Equal("say done", ((JsonArray)((JsonObject)stop[0]!)["hooks"]!)[0]!["command"]!.GetValue<string>());
        Assert.NotNull(((JsonObject)root["hooks"]!)["PreToolUse"]);
        Assert.True(HooksConfig.IsInstalled(twice, Url));
    }

    [Fact]
    public void InstalledIsFalseWhenAnyEventIsMissing()
    {
        var json = HooksConfig.Install("", Url);
        var root = Obj(json);
        ((JsonObject)root["hooks"]!).Remove("SessionEnd");
        Assert.False(HooksConfig.IsInstalled(root.ToJsonString(), Url));
        Assert.False(HooksConfig.IsInstalled(HooksConfig.Install("", "http://127.0.0.1:50000/hook"), Url));
    }

    [Fact]
    public void RemoveLeavesOtherHooksAndDropsEmptyContainers()
    {
        const string existing = """{ "hooks": { "Stop": [ { "hooks": [ { "type": "command", "command": "say done" } ] } ] }, "theme": "dark" }""";
        var installed = HooksConfig.Install(existing, Url);
        var removed = HooksConfig.Remove(installed, Url);
        var root = Obj(removed);
        Assert.Equal("dark", root["theme"]!.GetValue<string>());
        var hooks = (JsonObject)root["hooks"]!;
        Assert.Equal(["Stop"], hooks.Select(kv => kv.Key).ToList());
        Assert.Single((JsonArray)hooks["Stop"]!);
        Assert.False(HooksConfig.IsInstalled(removed, Url));

        var bare = HooksConfig.Remove(HooksConfig.Install("{}", Url), Url);
        Assert.Equal("{}", bare.Replace(" ", "").Replace("\n", "").Replace("\r", ""));
    }

    [Fact]
    public void RemoveOnlyTouchesOurUrl()
    {
        var other = HooksConfig.Install("", "http://127.0.0.1:50000/hook");
        var both = HooksConfig.Install(other, Url);
        var removed = HooksConfig.Remove(both, Url);
        Assert.True(HooksConfig.IsInstalled(removed, "http://127.0.0.1:50000/hook"));
        Assert.False(HooksConfig.IsInstalled(removed, Url));
    }

    [Fact]
    public void AcceptsCommentsAndTrailingCommasButRejectsGarbage()
    {
        var json = HooksConfig.Install("{ // user settings\n \"model\": \"opus\", }", Url);
        Assert.True(HooksConfig.IsInstalled(json, Url));
        Assert.Throws<JsonException>(() => HooksConfig.Install("{ nope", Url));
        Assert.Throws<JsonException>(() => HooksConfig.Install("[]", Url));
    }
}

File: mac/Tests/ClaudeToolbarCoreTests/Sessions/HooksConfigTests.swift
import XCTest
@testable import ClaudeToolbarCore

final class HooksConfigTests: XCTestCase {
    private let url = "http://127.0.0.1:47831/hook"

    private func object(_ json: String) throws -> [String: Any] {
        try XCTUnwrap(JSONSerialization.jsonObject(with: Data(json.utf8)) as? [String: Any])
    }

    private func groups(_ root: [String: Any], _ event: String) throws -> [[String: Any]] {
        let hooks = try XCTUnwrap(root["hooks"] as? [String: Any])
        return try XCTUnwrap(hooks[event] as? [[String: Any]])
    }

    func testUrlUsesThePort() {
        XCTAssertEqual(HooksConfig.hookUrl(port: 47831), url)
    }

    func testInstallIntoEmptyCreatesEveryEvent() throws {
        let json = try HooksConfig.install("", url: url)
        let root = try object(json)
        let hooks = try XCTUnwrap(root["hooks"] as? [String: Any])
        XCTAssertEqual(Set(hooks.keys), Set(HooksConfig.events))
        for event in HooksConfig.events {
            let handlers = try XCTUnwrap(try groups(root, event)[0]["hooks"] as? [[String: Any]])
            XCTAssertEqual(handlers[0]["type"] as? String, "http")
            XCTAssertEqual(handlers[0]["url"] as? String, url)
            XCTAssertEqual(handlers[0]["timeout"] as? Int, 5)
        }
        XCTAssertTrue(try HooksConfig.isInstalled(json, url: url))
        XCTAssertFalse(try HooksConfig.isInstalled("{}", url: url))
        XCTAssertFalse(try HooksConfig.isInstalled("", url: url))
    }

    func testInstallIsIdempotentAndPreservesOtherContent() throws {
        let existing = """
        { "model": "opus", "hooks": { "Stop": [ { "hooks": [ { "type": "command", "command": "say done" } ] } ],
          "PreToolUse": [ { "matcher": "Bash", "hooks": [ { "type": "command", "command": "lint" } ] } ] } }
        """
        let once = try HooksConfig.install(existing, url: url)
        let twice = try HooksConfig.install(once, url: url)
        XCTAssertEqual(once, twice)
        let root = try object(twice)
        XCTAssertEqual(root["model"] as? String, "opus")
        let stop = try groups(root, "Stop")
        XCTAssertEqual(stop.count, 2)
        XCTAssertEqual((stop[0]["hooks"] as? [[String: Any]])?[0]["command"] as? String, "say done")
        XCTAssertNotNil((root["hooks"] as? [String: Any])?["PreToolUse"])
        XCTAssertTrue(try HooksConfig.isInstalled(twice, url: url))
    }

    func testInstalledIsFalseWhenAnyEventIsMissing() throws {
        let other = try HooksConfig.install("", url: "http://127.0.0.1:50000/hook")
        XCTAssertFalse(try HooksConfig.isInstalled(other, url: url))
        let json = try HooksConfig.install("", url: url)
        var root = try object(json)
        var hooks = try XCTUnwrap(root["hooks"] as? [String: Any])
        hooks.removeValue(forKey: "SessionEnd")
        root["hooks"] = hooks
        let trimmed = String(decoding: try JSONSerialization.data(withJSONObject: root), as: UTF8.self)
        XCTAssertFalse(try HooksConfig.isInstalled(trimmed, url: url))
    }

    func testRemoveLeavesOtherHooksAndDropsEmptyContainers() throws {
        let existing = #"{ "hooks": { "Stop": [ { "hooks": [ { "type": "command", "command": "say done" } ] } ] }, "theme": "dark" }"#
        let installed = try HooksConfig.install(existing, url: url)
        let removed = try HooksConfig.remove(installed, url: url)
        let root = try object(removed)
        XCTAssertEqual(root["theme"] as? String, "dark")
        let hooks = try XCTUnwrap(root["hooks"] as? [String: Any])
        XCTAssertEqual(Array(hooks.keys), ["Stop"])
        XCTAssertEqual(try groups(root, "Stop").count, 1)
        XCTAssertFalse(try HooksConfig.isInstalled(removed, url: url))

        let bare = try HooksConfig.remove(try HooksConfig.install("{}", url: url), url: url)
        XCTAssertTrue(try object(bare).isEmpty)
    }

    func testRemoveOnlyTouchesOurUrl() throws {
        let other = "http://127.0.0.1:50000/hook"
        let both = try HooksConfig.install(try HooksConfig.install("", url: other), url: url)
        let removed = try HooksConfig.remove(both, url: url)
        XCTAssertTrue(try HooksConfig.isInstalled(removed, url: other))
        XCTAssertFalse(try HooksConfig.isInstalled(removed, url: url))
    }

    func testRejectsGarbage() {
        XCTAssertThrowsError(try HooksConfig.install("{ nope", url: url))
        XCTAssertThrowsError(try HooksConfig.install("[]", url: url))
        XCTAssertThrowsError(try HooksConfig.isInstalled("{ nope", url: url))
    }
}
